lessons/4.mdx

```diff
@@ -0,0 +1,818 @@
+  debug which piece actually broken.
+- Therefore it's best practice to have many separate tests with one `expect` in
+  each. In practice we could have tested both the name and the symbol in one
+  test, but see bullet 1 above!
+- And basically every test has two pieces (in addition to the `beforeEach()`
+  setup code): a call of a method to be tested, and code to see if the expected
+  result happened.
+
+Lastly when you write a test, always make sure it fails first! So try changing
+`expect` line to test the name `Anita` instead of `TierNFT`, run the test via
+`npm test` and make sure it fails. If your test passes when you make an obvious
+change like that, something is wrong with your test.
+
+So after intentionally causing one of our tests to fail, here is what the
+failure looks like:
+
+```bash
+
+  TierNFT
+    construction
+      1) set proper collection name
+      ✔ set proper collection symbol
+
+
+  1 passing (2s)
+  1 failing
+
+  1) TierNFT
+       construction
+         set proper collection name:
+
+      AssertionError: expected 'TierNFT' to equal 'Anita'
+      + expected - actual
+
+      -TierNFT
+      +Anita
+```
+
+### Let's add tests for mint()
+
+Here are tests we'll be adding:
+
+```javascript
+describe('mint', async () => {
+  it('should not mint if value is below the minimum tier', async function () {
+    await expect(
+      contract.mint({
+        value: hre.ethers.utils.parseEther('0.001'),
+      }),
+    ).to.be.revertedWith('Not enough value for the minimum Tier')
+    await expect(await contract.totalSupply()).to.equal(0)
+  })
+
+  it('should increase total supply', async function () {
+    await contract.mint({
+      value: hre.ethers.utils.parseEther('0.01'),
+    })
+    await expect(await contract.totalSupply()).to.equal(1)
+  })
+
+  it('should mint Tier 0', async function () {
+    await contract.mint({
+      value: hre.ethers.utils.parseEther('0.01'),
+    })
+    await expect(await contract.tokenTier(1)).to.equal(0)
+  })
+
+  it('should mint Tier 1', async function () {
+    await contract.mint({
+      value: hre.ethers.utils.parseEther('0.02'),
+    })
+    await expect(await contract.tokenTier(1)).to.equal(1)
+  })
+
+  it('should mint Tier 2', async function () {
+    await contract.mint({
+      value: hre.ethers.utils.parseEther('0.05'),
+    })
+    await expect(await contract.tokenTier(1)).to.equal(2)
+  })
+})
+```
+
+These tests introduce additional code that are helpful when testing contracts.
+
+With minting, one needs to pay the right price, we should also see the total
+supply of TierNFTs rise with each mint, and since we're minting 3 different
+tiers, we need to test each one of those gets minted based on the right price.
+
+The first test is a great example of how to test to ensure an expected error
+occurs. If not enough Eth is sent, the contract fails with an error and the
+transaction is reverted as if nothing happened.
+
+So the heavy-lifting happens in the
```

Comment:
again, here's a good opportunity to tell the new dev about heavy lifiting. If you could provide a quick scenario to contrast with **no heavy lifting** it might help. When I hear devs using this term, I'm usually a  bit lost, cos I never get to hear about the contrast

Reply:
True, this one is confusing, will fix. Going to also remove term heavy-lifting.